src/chapters/unit-testing/unit-testing-in-action.rst

```diff
@@ -0,0 +1,218 @@
+      assert.strictEqual(isPalindrome("ab"), false);
+   });
+
+   it("should be case-sensitive", function(){
+      assert.strictEqual(isPalindrome("abA"), false);
+   });
+
+   it("should consider whitespace", function(){
+      assert.strictEqual(isPalindrome("so many dynamos"), false);
+   });
+   
+Now run the tests to make sure they pass. Your code now includes a set of tests that considers a wide variety of positive and negative cases.
+
+Edge Cases
+----------
+
+Recall our definition of **edge case**:
+
+.. pull-quote::
+
+   An edge case is a test case that provides input at the extreme edge of what the unit should be able to handle.
+
+Edge cases can look very different for different units of code. Most of the examples we provided above dealt with numerical edge cases. However, edge cases can also be non-numeric. 
+
+In the case of ``isPalindrome``, the most obvious edge case would be that of the empty string, ``""``. This is the smallest possible string that we can use when calling ``isPalindrome``. Not only is it the smallest, but it is essentially *different* from the next longest string, ``"a"``---one has characters and one doesn't. 
+
+Should the empty string be considered a palindrome? That decision is up to us, the programmer, and there is not right or wrong answer. In our case, we have decided to take a very literal definition of the term "palindrome" by considering case and whitespace. In other words, our definition says that a string is a palindrome exactly when it equals its reverse. Since the reverse of ``""`` is also ``""``, it makes sense to consider the emtpy string a palindrome.
+
+Let's add this test case to our spec:
+
+.. sourcecode:: js
+
+   it("should consider the empty string a palindrome", function(){
+      assert.strictEqual(isPalindrome(""), true);
+   });
+
+Now run the tests, which should all pass.
+
+You might think that another edge case is that of the longest possible palindrome. Such a palindrome would be as long as the longest possible string in JavaScript. This case is not worth considering for a couple of reasons:
+
+- The length of the longest string `can vary across different JavaScript implementations <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length#Description>`_.
+- The most recent JavaScript specification, ES2016, states that the maximum allowed length of a string should be 2 :sup:`53` - 1 characters. This is a LOT of characters, and it is unrealistic to expect that our function will ever be given such a string.
+
+Toward a Better Testing Workflow
+--------------------------------
+
+In this case, we had a well-written function to write tests for, so it was straightforward to create tests that pass. Most situations will not be this simple. You will often find yourself writing tests for code only to discover a bug, forcing you to go back and update your code. That's okay! This is precisely what tests are for.
```

Comment:
Word removal and grammar fixes.
1st sentence: "In this case, we had a well-written function, so it was straightforward to create tests that passed."
3rd sentence: "Your tests will often uncover bugs in your code that need to be fixed."
Last sentence: "Do not panic! This is EXACTLY why we write tests."